Implement poutine.reparam with aux variable support

fritzo opened the pull request:
Addresses #2221 

This adds a `poutine.reparam` that transforms a sample site (based on `site["fn"]`) into one or more auxiliary sample sites followed by a deterministic transformation. This generalizes [(Gorinova et al. 2019)](https://arxiv.org/pdf/1906.03028.pdf) by allowing multiple "overparameterized" sample sites for auxiliary variable methods.

The main motivation for this PR is to support [Stable](http://docs.pyro.ai/en/stable/distributions.html#stable) distributions in variational inference. Although `Stable.log_prob()` is undefined, the distribution can be reparameterized to a pair of auxiliary variables with defined `.log_prob()`; thus `poutine.reparam` can make a stable model tractable.

I have designed to allow custom `Reparameterizer` instances to be set per-site, either manually or via `@infer_config`. This PR implements only `LocScaleReparameterizer` and `StableReparameterizer`. 

Note the `Reparameterizer` class restricts the dependency structure, disallowing one auxiliary variable to depend on another. In principle we could replace a `pyro.sample` site with an arbitrary probabilistic program (and the proposed `pyro.reparam()` interface could be adapted to this general case via an `isinstance(-,Reparameterizer)` check). One advantage of an opinionated `Reparameterizer` class is that if forces separation of concerns between `pyro.distributions` and `pyro.poutine`: no pyro primitives need be used in `Reparameterizer` implementations.

## Tested
- [x] test `TrivialReparameterizer`
- [x] test `LocScaleReparameterizer`
- [x] test `StableReparameterizer`
- [x] smoke test for `Stable` inference with autoguides
- [x] test SVI in tests/infer/test_valid_models.py
- [x] test MHC and NUTS in tests/infer/mcmc/test_valid_models.py
- [x] ensure `StableReparameterizer` is continuous

See generated [docs](http://fritzo.org/temp/pyro/html/distributions.html#reparameterizers)

fehiepsi commented on tests/distributions/test_reparameterize.py:
```diff
@@ -0,0 +1,107 @@
+    def config_fn(site):
+        if site["name"] == "x" and isinstance(site["fn"], dist.Normal):
+            return {"reparam": Reparam()}
+        return {}
```
I guess we don't need to return an empty dict here?

fritzo replied:
A dict is required by `infer_config`'s `config_fn`

fehiepsi commented:
In the future, it would be nice to have `TransformedReparameterizer` which handles transformed distribution. I would be very happy to use this pattern. :)
```
reparam_model = poutine.reparam(model, {"x": TransformedReparameterizer()})
```

neerajprad commented on pyro/poutine/reparam_messenger.py:
```diff
@@ -0,0 +1,61 @@
+            # Validate while the original msg["fn"] is known.
+            msg["fn"]._validate_sample(value)
+        msg["value"] = value
+        msg["fn"] = Delta(value, event_dim=msg["fn"].event_dim).mask(False)
```
Not directly relevant to this PR, but I think this will probably not work for HMC. In that case, we need to be able to score samples from a base distribution (like Normal(0., 1.)) using the transformed distribution's log density. The samples themselves will be generated by HMC, and the delta function (IIUC) will result in the sample getting rejected because of having infinite energy. Just an FYI; I haven't thought fully about it, but it may be the case that both are sufficiently different to warrant separate solutions. It is also straightforward to do non-centered parametrization manually, so we can ignore the MCMC case for the time being.

fritzo replied:
This is just a `pyro.deterministic()` call in-lined to avoid cyclic dependencies. Does HMC not yet support `pyro.deterministi()`?

fehiepsi replied:
I think this will work for HMC.

neerajprad replied:
Thanks for clarifying, I misread what this was doing. `pyro.deterministic` should work fine.

fritzo commented:
@fehiepsi I like the `TransformedParameterizer` idea! Especially because that broadens our class of transforms to non-invertible transforms: reparameterization avoids the need to implement `.log_abs_det_jacobian()`.

fehiepsi commented:
Yup, it is especially useful for flows with no `inverse` method (e.g. BNAFlow) and will make autoguide faster because, as you pointed out, no need to go through `transform.log_abs_det_jacobian`.

fritzo commented on pyro/distributions/reparameterize.py:
```diff
@@ -0,0 +1,78 @@
+class LocScaleReparameterizer(Reparameterizer):
+    """
+    Generic centering reparameterizer for distributions that are completely
+    spacified by parameters ``loc`` and ``scale``.
```
done

fritzo commented:
@fehiepsi I have updated to support both notations:
```py
# Either specify in the reparam config:
@poutine.reparam(config={"x": LocScaleReparameterizer()})
...
# or mark a single site.
pyro.sample("y", dist.Normal(loc, scale),
            infer={"reparam": LocScaleReparameterizer})
```
That way I can still use `@config_infer` to match sites with `isinstance(site["fn"], dist.Stable)`). 

@neerajprad I have added smoke tests for `HMC` and `NUTS`.

fritzo commented on pyro/poutine/reparam_messenger.py:
```diff
@@ -0,0 +1,72 @@
+            config = {}
+        self.config = config
+
+    def _pyro_sample(self, msg):
```
@eb8680 can you take a look to ensure I've implemented this correctly? I've followed `poutine.do`.

eb8680 replied:
I'll need to read the paper to understand so probably won't be able to get to this today

fritzo replied:
Oh I'm not so concerned with matching up with the paper as ensuring that I have correctly transformed a single `sample` site into multiple `sample` sites plus a `pyro.deterministic` site.

fritzo commented:
Hmm I may be missing a `log_abs_det_jacobian` term in `.transform_values()`...

fehiepsi commented:
I think we don't need `log_abs_det_jacobian` in autoguide/mcmc because the inference happens in unconstrained space (assume base dists have real support, this means that we only need to compute `p(x)` rather than `p(t(x))` where `x` belongs to base dists' domain and `t` is the transform).
